### ISSUE TITLE
check protobuf 

### ISSUE THREAD

**Issue by umarmanj1010** [DevLARLEY/mpdl#6]
![제목 없음](https://github.com/DevLARLEY/mpdl/assets/89239688/2845ef9d-cbe0-4dfa-bb4f-bf739f471e48)

what should i do?

**Comment by DevLARLEY** [comment]
I'm on vacation. I'll get back to you.

**Comment by umarmanj1010** [comment]
Enjoy your vacation and thank you for the great program

**Comment by DevLARLEY** [comment]
Could you please re-send the output as text?

**Comment by nihilistic-french-fries** [comment]
@DevLARLEY I too have recently encountered this error (with a fresh CDM). Here's the output in text form:
```
Worker(
https://archive-oregon.kiswe.com/media/kiswe_hangt_65061_2024_05_10_23_00_16_346/mix-stream1.mpd,
AAAAZnBzc2gAAAAA7e+LqXnWSs6jyCfc1R0h7QAAAEYIARIQoUgPvGPFSIyazITtTJXhgyIwZXlKaGMzTmxkRWxrSWpvaWNISnZaQzAyTlRBMk1TMXRhWGd0YzNSeVpXRnRNU0o5,
https://lic46.drmtoday.com/license-proxy-widevine/cenc/,
f3f6c353-b3dd-4da3-9ad1-c40528a24763)
unable to parse license - check protobufs
```

**Comment by DevLARLEY** [comment]
The reason this is happening is because the license server requires the challenge to be in a specific format. I'll soon add more formatting schemes so MPDL just uses trial-and-error until it gets a proper response.

**Comment by DevLARLEY** [comment]
What site are you streaming from?

**Comment by nihilistic-french-fries** [comment]
It's a VOD from https://live.smosh.com/events/sitcom

**Comment by DevLARLEY** [comment]
I currently cannot reproduce the key request. Could you send the request and response of the license server and the request headers from developer tools?

**Comment by nihilistic-french-fries** [comment]
I attached the POST request and response below. There was a second POST request made with a larger payload; I removed the curl command's `--data-raw` argument since it was encoded but figured the rest of the request's headers would be valuable.

**First POST Request**

Request:
```bash
curl 'https://lic46.drmtoday.com/license-proxy-widevine/cenc/' \
  -H 'accept: */*' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'dnt: 1' \
  -H 'dt-custom-data;' \
  -H 'origin: https://live.smosh.com' \
  -H 'pragma: no-cache' \
  -H 'priority: u=1, i' \
  -H 'referer: https://live.smosh.com/' \
  -H 'sec-ch-ua: "Chromium";v="124", "Google Chrome";v="124", "Not-A.Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: cross-site' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36' \
  -H 'x-dt-auth-token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9.eyJvcHREYXRhIjoie1widXNlcklkXCI6IFwiMTc3Mzk5NzlcIiwgXCJtZXJjaGFudFwiOiBcImtpc3dlXCIsIFwic2Vzc2lvbklkXCI6IFwiUFJPRC4xNzczOTk3OUAyMDI0LTA1LTEyIDAwOjMzOjEyLjU0NjYwMyswMDowMFwifSIsImNydCI6Ilt7XCJhc3NldElkXCI6IFwicHJvZC02NTA2Mi1taXgtc3RyZWFtMVwiLCBcImFjY291bnRpbmdJZFwiOiBcInBhaWRfZXZlbnQuMjYyLnN0cmVhbS5taXgtc3RyZWFtMS50eXBlLmRhc2hcIiwgXCJwcm9maWxlXCI6IHtcInB1cmNoYXNlXCI6IHt9fX1dIiwiaWF0IjoxNzE1NDczOTkyLCJqdGkiOiJiMzkxODgwNC05NTY3LTQ5NmItYWEyZC1hYTMzZDM5NzdiNjAifQ.hPB3C8T_5trAi6suWZMGgLskr6hdJdYkch9NuajKIdpybvGu26sM6Ckl-eVbqOiqt-jS3ii_KUXegaCKTGD47A' \
  -H 'x-dt-custom-data: eyJ1c2VySWQiOiAiMTc3Mzk5NzkiLCAibWVyY2hhbnQiOiAia2lzd2UiLCAic2Vzc2lvbklkIjogIlBST0QuMTc3Mzk5NzlAMjAyNC0wNS0xMiAwMDozMzoxMi41NDY2MDMrMDA6MDAifQ==' \
  --data-raw $'\u0008\u0004'
```

Response:
```json
{
    "status": "OK",
    "license": "CAUSwQUKuwIIAxIQoNzRYDAsBBPVIHaiDIEkTBi/hPHuBSKOAjCCAQoCggEBALxdz8ZNHBkO2WcFOGrxXfzYPWgkAzQX9iAJC00MvPu9NYn5CyQ0aLkN3arM7izjwKfHkEu8vgfTEJFLDYR3p6QdJ2tFTDKhg4T4z+3WghtxzCVNDuj4GlXVfpMHt/EXLHWOwG1GwFamFTrnDXbXOqZKLDNnMnjGxdVAImlCc0jrWDMc7jZdk4y7HK0ybQXqm40dNOqHJjdnb3Riq9IN04H5RjlAm5cmCOG85QP36houW9P6/CZXb80lbEqocEp0v0mq86c2JBQXSEMjtZ/xZvrdFhJcoTbsrLfWEqDHXC6rtNNaQD6QeclNStMR9DF89Hh0lOq4T4gMsCLnJ5VXHNsCAwEAAToMY2FzdGxhYnMuY29tQAESgAMyYZeleD/SrFS7CiSmypUgdmsl+7rZqv4EXxRVQp0rmhhCiJNl307MreSvb2SoYaTeVKAzT+nj32n8VQvygcfOCwnsdqN2zLOfRWzf5XKwCBokCRdn6T43oprRYGJmUK1y6dUkOIdBvPWwX0a9x49RSLZ8z+HuHKyQblksDIkzNO+CdXcPz8jljqYIUE9v/79GL71sS+q48XcV/2KnfIQe72ZxB4RgJ4Z73NH/bt4jkwBNff2Bi0Keft0NdXGjj+sJgrQDhV6YLcdFqLZfelIZywpjv7rB7uHgjJyOq24a1iDwBtqKCi9yhREBlyvV38bSulESMt1UsbQaOqTrcbHrzxbmCHdMp6awrdHWNXR05+Pd7q5X+ROmw1Ev0qwPS/vKwFBV1RcRsmi1wXPvW0jmpzePCvoaNgg8DdXO6wPXs83oAZCeIzDElXXbP8SCdSilwUHXeizJ2kPYBoOyBajlFYlw1/5/MKC1RDH/QC737FWbktAF2/wlrh/aiIc3d3A=",
    "service_version_info":
    {
        "license_sdk_version": "18.1.2",
        "license_service_version": "DRMtoday"
    }
}
```

**Second POST Request**

Request:
```bash
curl 'https://lic46.drmtoday.com/license-proxy-widevine/cenc/' \
  -H 'accept: */*' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'dnt: 1' \
  -H 'dt-custom-data;' \
  -H 'origin: https://live.smosh.com' \
  -H 'pragma: no-cache' \
  -H 'priority: u=1, i' \
  -H 'referer: https://live.smosh.com/' \
  -H 'sec-ch-ua: "Chromium";v="124", "Google Chrome";v="124", "Not-A.Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: cross-site' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36' \
  -H 'x-dt-auth-token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9.eyJvcHREYXRhIjoie1widXNlcklkXCI6IFwiMTc3Mzk5NzlcIiwgXCJtZXJjaGFudFwiOiBcImtpc3dlXCIsIFwic2Vzc2lvbklkXCI6IFwiUFJPRC4xNzczOTk3OUAyMDI0LTA1LTEyIDAwOjMzOjEyLjU0NjYwMyswMDowMFwifSIsImNydCI6Ilt7XCJhc3NldElkXCI6IFwicHJvZC02NTA2Mi1taXgtc3RyZWFtMVwiLCBcImFjY291bnRpbmdJZFwiOiBcInBhaWRfZXZlbnQuMjYyLnN0cmVhbS5taXgtc3RyZWFtMS50eXBlLmRhc2hcIiwgXCJwcm9maWxlXCI6IHtcInB1cmNoYXNlXCI6IHt9fX1dIiwiaWF0IjoxNzE1NDczOTkyLCJqdGkiOiJiMzkxODgwNC05NTY3LTQ5NmItYWEyZC1hYTMzZDM5NzdiNjAifQ.hPB3C8T_5trAi6suWZMGgLskr6hdJdYkch9NuajKIdpybvGu26sM6Ckl-eVbqOiqt-jS3ii_KUXegaCKTGD47A' \
  -H 'x-dt-custom-data: eyJ1c2VySWQiOiAiMTc3Mzk5NzkiLCAibWVyY2hhbnQiOiAia2lzd2UiLCAic2Vzc2lvbklkIjogIlBST0QuMTc3Mzk5NzlAMjAyNC0wNS0xMiAwMDozMzoxMi41NDY2MDMrMDA6MDAifQ=='
```

**Comment by nihilistic-french-fries** [comment]
I also noticed a GET request to another URL that seems to be returning the contents of the `x-dt-auth-token` & `x-dt-custom-data` headers found in the drmtoday.com POST requests. Request and response below:

Request:
```bash
curl 'https://thor-api-region-central.kiswe.com/thor/video/drm/?video=cb50725b-c7f4-4f9b-a4f5-aa0bf6fa205f&stream=mix-stream1&paid_event=262&stream_type=dash' \
  -H 'accept: application/json; app=kiswe; version=1.10.6; platform=web; os_version=Chrome 124.0.0.0; device_type=Windows 10 undefined' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'authorization: Token 149b43bfb4cf74c76d8a1336fc72d07fec361c84' \
  -H 'cache-control: no-cache' \
  -H 'dnt: 1' \
  -H 'origin: https://live.smosh.com' \
  -H 'pragma: no-cache' \
  -H 'priority: u=1, i' \
  -H 'referer: https://live.smosh.com/' \
  -H 'sec-ch-ua: "Chromium";v="124", "Google Chrome";v="124", "Not-A.Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: cross-site' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36'
```

Response;
```json
{
    "x-dt-auth-token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9.eyJvcHREYXRhIjoie1widXNlcklkXCI6IFwiMTc3Mzk5NzlcIiwgXCJtZXJjaGFudFwiOiBcImtpc3dlXCIsIFwic2Vzc2lvbklkXCI6IFwiUFJPRC4xNzczOTk3OUAyMDI0LTA1LTEyIDAxOjExOjIzLjUzNDIyNCswMDowMFwifSIsImNydCI6Ilt7XCJhc3NldElkXCI6IFwicHJvZC02NTA2Mi1taXgtc3RyZWFtMVwiLCBcImFjY291bnRpbmdJZFwiOiBcInBhaWRfZXZlbnQuMjYyLnN0cmVhbS5taXgtc3RyZWFtMS50eXBlLmRhc2hcIiwgXCJwcm9maWxlXCI6IHtcInB1cmNoYXNlXCI6IHt9fX1dIiwiaWF0IjoxNzE1NDc2MjgzLCJqdGkiOiI1Y2ZiY2I3Mi0zN2FjLTRjM2MtOTI3MC1iN2M4YjZkNmVlYjYifQ.cQ0fRFOyYlKwPeIdq_ucnX7GiXdhj_OZNPfUnVqMNpNkBMgyTCmFrSf8K_xzYLXKA_F4VP_z0j9uglGPmCb4gw",
    "x-dt-custom-data": "eyJ1c2VySWQiOiAiMTc3Mzk5NzkiLCAibWVyY2hhbnQiOiAia2lzd2UiLCAic2Vzc2lvbklkIjogIlBST0QuMTc3Mzk5NzlAMjAyNC0wNS0xMiAwMToxMToyMy41MzQyMjQrMDA6MDAifQ=="
}
```

**Comment by DevLARLEY** [comment]
I don't know why, but i can't even get a different DRMToday service to work right now (VTMGo.be) manually (a pywidevine script).
(Getting a 400 - Bad Request)
The successor to mpdl that I'm working on is the only program that currently works with VTMGo for me. Even though it's site-specifc, it's going to be really easy to add support for a whole streaming service even if you're not a professional programmer.
MPDL may be deprecated after the release.

**Comment by nihilistic-french-fries** [comment]
No worries, there may be some IP or region allow-listing they're doing on the backend. I appreciate your assistance and the wonderful tooling you've crafted!

In the end I did manage to get it working in a bespoke script that accounted for the service certificate retrieval that is used to sign the license challenge request (these were the two license server POST requests we saw previously).
